processing/src/main/java/org/apache/druid/query/dimension/PrefixFilteredDimensionSpec.java

```diff
@@ -0,0 +1,144 @@
+
+  public PrefixFilteredDimensionSpec(
+      @JsonProperty("delegate") DimensionSpec delegate,
+      @JsonProperty("prefix") String prefix //rows not starting with the prefix will be discarded
+  )
+  {
+    super(delegate);
+    this.prefix = Preconditions.checkNotNull(prefix, "prefix must not be null");
+  }
+
+  @JsonProperty
+  public String getPrefix()
+  {
+    return prefix;
+  }
+
+  @Override
+  public DimensionSelector decorate(final DimensionSelector selector)
+  {
+    if (selector == null) {
+      return null;
+    }
+
+    final int selectorCardinality = selector.getValueCardinality();
+    if (selectorCardinality < 0 || !selector.nameLookupPossibleInAdvance()) {
+      return new PredicateFilteredDimensionSelector(
+          selector,
+          new Predicate<String>()
+          {
+            @Override
+            public boolean apply(@Nullable String input)
+            {
+              return NullHandling.nullToEmptyIfNeeded(input).startsWith(prefix);
+            }
+          }
+      );
+    }
+
+    int count = 0;
+    final Int2IntOpenHashMap forwardMapping = new Int2IntOpenHashMap();
+    forwardMapping.defaultReturnValue(-1);
+    for (int i = 0; i < selectorCardinality; i++) {
+      if (NullHandling.nullToEmptyIfNeeded(selector.lookupName(i)).startsWith(prefix)) {
+        forwardMapping.put(i, count++);
+      }
+    }
+
+    final int[] reverseMapping = new int[forwardMapping.size()];
+    for (Int2IntMap.Entry e : forwardMapping.int2IntEntrySet()) {
+      reverseMapping[e.getIntValue()] = e.getIntKey();
+    }
+    return new ForwardingFilteredDimensionSelector(selector, forwardMapping, reverseMapping);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] delegateCacheKey = delegate.getCacheKey();
+    byte[] prefixBytes = StringUtils.toUtf8(NullHandling.nullToEmptyIfNeeded(prefix));
+    return ByteBuffer.allocate(2 + delegateCacheKey.length + prefixBytes.length)
+                     .put(CACHE_TYPE_ID)
+                     .put(delegateCacheKey)
+                     .put(DimFilterUtils.STRING_SEPARATOR)
+                     .put(prefixBytes)
+                     .array();
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    PrefixFilteredDimensionSpec that = (PrefixFilteredDimensionSpec) o;
+
+    if (!delegate.equals(that.delegate)) {
+      return false;
+    }
+    return NullHandling.nullToEmptyIfNeeded(prefix)
+        .equals(NullHandling.nullToEmptyIfNeeded(that.prefix));
+
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int result = delegate.hashCode();
+    result = 31 * result + NullHandling.nullToEmptyIfNeeded(prefix).hashCode();
+    return result;
+  }
+
+  @Override
+  public String toString()
+  {
+    return "PrefixFilteredDimensionSpec{" +
+           "Prefix='" + NullHandling.nullToEmptyIfNeeded(prefix) + '\'' +
```

Comment:
"prefix" has a null check in the constructor, so the nullToEmpty call is not needed here